api/options.go

```diff
@@ -0,0 +1,110 @@
+package api
+
+import (
+	"context"
+	"net/http"
+
+	dperrors "github.com/ONSdigital/dp-cantabular-filter-flex-api/errors"
+	"github.com/ONSdigital/dp-cantabular-filter-flex-api/model"
+	"github.com/go-chi/chi/v5"
+	"github.com/pkg/errors"
+)
+
+func (api *API) getFilterDimensionOptions(w http.ResponseWriter, r *http.Request) {
+	ctx := context.Background()
+
+	filterID := chi.URLParam(r, "id")
+	dimensionName := chi.URLParam(r, "name")
+
+	pageLimit, offset, err := getPaginationParams(r.URL, api.cfg.DefaultMaximumLimit)
+	if err != nil {
+		api.respond.Error(
+			ctx,
+			w,
+			http.StatusBadRequest,
+			errors.Wrap(err, "Bad Request: "),
+		)
+
+	}
+
+	if pageLimit == 0 {
+		// define a reasonable default
+		// in light of bad input
+		// also slice not work with 0
+		pageLimit = 10
+	}
+
+	var eTag string
+	if reqETag := api.getETag(r); reqETag != eTagAny {
+		eTag = reqETag
+	}
+
+	options, totalCount, err := api.store.GetFilterDimensionOptions(
+		ctx,
+		filterID,
+		dimensionName,
+		pageLimit,
+		offset,
+	)
+	if err != nil {
+		status := statusCode(err)
+		finalErr := errors.New("internal server error")
+
+		if dperrors.NotFound(err) {
+			status = http.StatusNotFound
+			finalErr = err
+		}
+		if dperrors.BadRequest(err) {
+			status = http.StatusBadRequest
+			finalErr = err
+		}
+
+		api.respond.Error(
+			ctx,
+			w,
+			status,
+			errors.Wrap(finalErr, "Failed to get options"),
+		)
+		return
+	}
+
+	resp := GetDimensionOptionsResponse{
+		Items: parseFilterDimensionOptions(options, filterID, dimensionName),
+		paginationResponse: paginationResponse{
+			Limit:      pageLimit,
+			Offset:     offset,
+			Count:      len(options),
+			TotalCount: totalCount,
+		},
+	}
+
+	w.Header().Set(eTagHeader, eTag)
```

Comment:
I am not sure that I understand fully the process of retrieving and setting this eTag. I would like some guidance on this.

Reply:
The whole point of the eTags is that the caller will send an eTag in the request.. You then need to check that it matches with the stored eTag record. If it doesn't you need to return http.StatusConflict. When updating a record you need to generate a new eTag, store it and return it in the header. What you've done here is just return the eTag the caller has sent which is wrong. It looks like some endpoints have missed implementing the eTags in general but we really should be doing it. We should probably add a ticket to go back through each of them and make sure they're all correct.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

Reply:
I have ammended the mongo query to retrieve the etag and add to the header